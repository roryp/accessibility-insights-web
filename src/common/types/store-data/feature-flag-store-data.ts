// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { DictionaryStringTo } from 'types/common-types';

export type FeatureFlagStoreData = DictionaryStringTo<boolean>;
