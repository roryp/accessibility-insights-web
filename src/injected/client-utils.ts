// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
export interface ClientRectOffset {
    left: number;
    top: number;
}

export interface ScrollAccessor {
    scrollX: number;
    scrollY: number;
}

export interface BoundRectAccessor {
    getBoundingClientRect: () => ClientRectOffset;
}

export interface ElementMatcher {
    matches?: (selector: string) => boolean;
    webkitMatchesSelector?: (selector: string) => boolean;
    msMatchesSelector?: (selector: string) => boolean;
}

export class ClientUtils {
    private scroll: ScrollAccessor;

    constructor(scroll: ScrollAccessor) {
        this.scroll = scroll;
    }

    public getOffset(element: BoundRectAccessor): ClientRectOffset {
        const elementRect = element.getBoundingClientRect();

        return this.getOffsetFromBoundingRect(elementRect);
    }

    public getOffsetFromBoundingRect(elementRect: ClientRect | ClientRectOffset): ClientRectOffset {
        return {
            left: elementRect.left + this.scroll.scrollX,
            top: elementRect.top + this.scroll.scrollY,
        };
    }

    public matchesSelector(element: ElementMatcher, selectorString: string): boolean {
        const selector = (
            element.matches ||
            element.webkitMatchesSelector ||
            element.msMatchesSelector
        ).bind(element);

        return selector(selectorString);
    }
}
