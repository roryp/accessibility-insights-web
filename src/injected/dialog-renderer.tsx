// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { getRTL } from '@uifabric/utilities';
import { IssueDetailsTextGenerator } from 'background/issue-details-text-generator';
import { NavigatorUtils } from 'common/navigator-utils';
import * as React from 'react';
import * as ReactDOM from 'react-dom';
import { BrowserAdapter } from '../common/browser-adapters/browser-adapter';
import { FixInstructionProcessor } from '../common/components/fix-instruction-processor';
import { NewTabLink } from '../common/components/new-tab-link';
import { FeatureFlags } from '../common/feature-flags';
import { HTMLElementUtils } from '../common/html-element-utils';
import { getPlatform } from '../common/platform';
import { FeatureFlagStoreData } from '../common/types/store-data/feature-flag-store-data';
import { WindowUtils } from '../common/window-utils';
import { createIssueDetailsBuilder } from '../issue-filing/common/create-issue-details-builder';
import { IssueFilingUrlStringUtils } from '../issue-filing/common/issue-filing-url-string-utils';
import { PlainTextFormatter } from '../issue-filing/common/markup/plain-text-formatter';
import { AxeResultToIssueFilingDataConverter } from '../issue-filing/rule-result-to-issue-filing-data';
import { DictionaryStringTo } from '../types/common-types';
import { rootContainerId } from './constants';
import { DetailsDialogHandler } from './details-dialog-handler';
import { ErrorMessageContent } from './frameCommunicators/error-message-content';
import { FrameCommunicator, MessageRequest } from './frameCommunicators/frame-communicator';
import { FrameMessageResponseCallback } from './frameCommunicators/window-message-handler';
import {
    LayeredDetailsDialogComponent,
    LayeredDetailsDialogDeps,
} from './layered-details-dialog-component';
import { MainWindowContext } from './main-window-context';
import { DecoratedAxeNodeResult, HtmlElementAxeResults } from './scanner-utils';
import { ShadowUtils } from './shadow-utils';

export interface DetailsDialogWindowMessage {
    data: HtmlElementAxeResults;
    featureFlagStoreData: FeatureFlagStoreData;
}

export type RenderDialog = (
    data: HtmlElementAxeResults,
    featureFlagStoreData: FeatureFlagStoreData,
) => void;

export class DialogRenderer {
    private static readonly renderDetailsDialogCommand = 'insights.detailsDialog';

    constructor(
        private readonly dom: Document,
        private readonly renderer: typeof ReactDOM.render,
        private readonly frameCommunicator: FrameCommunicator,
        private readonly htmlElementUtils: HTMLElementUtils,
        private readonly windowUtils: WindowUtils,
        private readonly navigatorUtils: NavigatorUtils,
        private readonly shadowUtils: ShadowUtils,
        private readonly browserAdapter: BrowserAdapter,
        private readonly getRTLFunc: typeof getRTL,
        private readonly detailsDialogHandler: DetailsDialogHandler,
    ) {
        if (this.isInMainWindow()) {
            this.frameCommunicator.subscribe(
                DialogRenderer.renderDetailsDialogCommand,
                this.processRequest,
            );
        }
    }

    public render: RenderDialog = (
        data: HtmlElementAxeResults,
        featureFlagStoreData: FeatureFlagStoreData,
    ) => {
        if (this.isInMainWindow()) {
            const mainWindowContext = MainWindowContext.getMainWindowContext();
            mainWindowContext.getTargetPageActionMessageCreator().openIssuesDialog();

            const elementSelector: string = this.getElementSelector(data);
            const failedRules: DictionaryStringTo<DecoratedAxeNodeResult> = this.getFailedRules(
                data,
            );
            const target: string[] = this.getTarget(data);
            const dialogContainer: HTMLDivElement = featureFlagStoreData[FeatureFlags.shadowDialog]
                ? this.initializeDialogContainerInShadowDom()
                : this.appendDialogContainer();

            const issueDetailsTextGenerator = new IssueDetailsTextGenerator(
                IssueFilingUrlStringUtils,
                createIssueDetailsBuilder(PlainTextFormatter),
            );

            const fixInstructionProcessor = new FixInstructionProcessor();

            const axeResultToIssueFilingDataConverter = new AxeResultToIssueFilingDataConverter(
                IssueFilingUrlStringUtils.getSelectorLastPart,
            );

            const deps: LayeredDetailsDialogDeps = {
                axeResultToIssueFilingDataConverter,
                fixInstructionProcessor,
                issueDetailsTextGenerator,
                windowUtils: this.windowUtils,
                navigatorUtils: this.navigatorUtils,
                targetPageActionMessageCreator: mainWindowContext.getTargetPageActionMessageCreator(),
                issueFilingActionMessageCreator: mainWindowContext.getIssueFilingActionMessageCreator(),
                browserAdapter: this.browserAdapter,
                getRTL: this.getRTLFunc,
                toolData: mainWindowContext.getToolData(),
                issueFilingServiceProvider: mainWindowContext.getIssueFilingServiceProvider(),
                userConfigMessageCreator: mainWindowContext.getUserConfigMessageCreator(),
                LinkComponent: NewTabLink,
            };

            this.renderer(
                <LayeredDetailsDialogComponent
                    deps={deps}
                    failedRules={failedRules}
                    elementSelector={elementSelector}
                    target={target}
                    dialogHandler={this.detailsDialogHandler}
                    devToolStore={mainWindowContext.getDevToolStore()}
                    userConfigStore={mainWindowContext.getUserConfigStore()}
                    devToolsShortcut={getPlatform(this.windowUtils).devToolsShortcut}
                    devToolActionMessageCreator={mainWindowContext.getDevToolActionMessageCreator()}
                    featureFlagStoreData={featureFlagStoreData}
                />,
                dialogContainer,
            );
        } else {
            const windowMessageRequest: MessageRequest<DetailsDialogWindowMessage> = {
                win: this.windowUtils.getTopWindow(),
                command: DialogRenderer.renderDetailsDialogCommand,
                message: { data: data, featureFlagStoreData: featureFlagStoreData },
            };
            this.frameCommunicator.sendMessage(windowMessageRequest);
        }
    };

    private processRequest = (
        message: DetailsDialogWindowMessage,
        error: ErrorMessageContent,
        sourceWin: Window,
        responder?: FrameMessageResponseCallback,
    ): void => {
        this.render(message.data, message.featureFlagStoreData);
    };

    private initializeDialogContainerInShadowDom(): HTMLDivElement {
        const shadowContainer = this.shadowUtils.getShadowContainer();

        const dialogContainer = this.dom.createElement('div');
        dialogContainer.className = 'insights-shadow-dialog-container';
        shadowContainer.appendChild(dialogContainer);
        return dialogContainer;
    }

    private appendDialogContainer(): HTMLDivElement {
        this.htmlElementUtils.deleteAllElements('.insights-dialog-container');

        const dialogContainer = this.dom.createElement('div');
        dialogContainer.setAttribute('class', 'insights-dialog-container');
        this.dom.querySelector(`#${rootContainerId}`).appendChild(dialogContainer);
        return dialogContainer;
    }

    private getFailedRules(
        data: HtmlElementAxeResults,
    ): DictionaryStringTo<DecoratedAxeNodeResult> {
        return data.ruleResults;
    }

    private getTarget(data: HtmlElementAxeResults): string[] {
        return data.target;
    }

    private getElementSelector(data: HtmlElementAxeResults): string {
        return data.target.join(';');
    }

    private isInMainWindow(): boolean {
        return this.windowUtils.getTopWindow() === this.windowUtils.getWindow();
    }
}
