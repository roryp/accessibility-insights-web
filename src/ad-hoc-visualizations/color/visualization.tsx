// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { AdHocTestkeys } from 'common/configs/adhoc-test-keys';
import { TestMode } from 'common/configs/test-mode';
import { VisualizationConfiguration } from 'common/configs/visualization-configuration';
import { Messages } from 'common/messages';
import { TelemetryDataFactory } from 'common/telemetry-data-factory';
import { VisualizationType } from 'common/types/visualization-type';
import { generateUID } from 'common/uid-generator';
import { adhoc as content } from 'content/adhoc';
import { AdhocStaticTestView } from 'DetailsView/components/adhoc-static-test-view';
import { RuleAnalyzerConfiguration } from 'injected/analyzers/analyzer';
import { ScannerUtils } from 'injected/scanner-utils';
import { VisualizationInstanceProcessor } from 'injected/visualization-instance-processor';
import * as React from 'react';

const { guidance } = content.color;

const colorRuleAnalyzerConfiguration: RuleAnalyzerConfiguration = {
    rules: ['select-body'],
    resultProcessor: (scanner: ScannerUtils) => scanner.getAllCompletedInstances,
    telemetryProcessor: (telemetryFactory: TelemetryDataFactory) => telemetryFactory.forTestScan,
    key: AdHocTestkeys.Color,
    testType: VisualizationType.Color,
    analyzerMessageType: Messages.Visualizations.Common.ScanCompleted,
};

export const ColorAdHocVisualization: VisualizationConfiguration = {
    getTestView: props => <AdhocStaticTestView {...props} />,
    key: AdHocTestkeys.Color,
    testMode: TestMode.Adhoc,
    getStoreData: data => data.adhoc.color,
    enableTest: (data, _) => (data.enabled = true),
    disableTest: data => (data.enabled = false),
    getTestStatus: data => data.enabled,
    displayableData: {
        title: 'Color',
        enableMessage: 'Changing color to greyscale...',
        toggleLabel: 'Show grayscale',
        linkToDetailsViewText: 'How to test color',
    },
    chromeCommand: '05_toggle-color',
    launchPanelDisplayOrder: 5,
    adhocToolsPanelDisplayOrder: 2,
    getAnalyzer: provider => provider.createRuleAnalyzer(colorRuleAnalyzerConfiguration),
    getIdentifier: () => AdHocTestkeys.Color,
    visualizationInstanceProcessor: () => VisualizationInstanceProcessor.nullProcessor,
    getNotificationMessage: selectorMap => null,
    getDrawer: provider => provider.createSingleTargetDrawer('insights-grey-scale-container'),
    getSwitchToTargetTabOnScan: () => false,
    getInstanceIdentiferGenerator: () => generateUID,
    guidance,
};
