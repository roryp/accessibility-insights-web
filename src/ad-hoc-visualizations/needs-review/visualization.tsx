// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { NeedsReviewInstancesSection } from 'common/components/cards/needs-review-instances-section';
import { RuleAnalyzerConfiguration } from 'injected/analyzers/analyzer';
import * as React from 'react';
import { getNotificationMessage } from '../../ad-hoc-visualizations/issues/get-notification-message';
import { AdHocTestkeys } from '../../common/configs/adhoc-test-keys';
import { TestMode } from '../../common/configs/test-mode';
import { VisualizationConfiguration } from '../../common/configs/visualization-configuration';
import { FeatureFlags } from '../../common/feature-flags';
import { Messages } from '../../common/messages';
import { TelemetryDataFactory } from '../../common/telemetry-data-factory';
import { VisualizationType } from '../../common/types/visualization-type';
import { generateUID } from '../../common/uid-generator';
import { AdhocIssuesTestView } from '../../DetailsView/components/adhoc-issues-test-view';
import { ScannerUtils } from '../../injected/scanner-utils';
import { VisualizationInstanceProcessor } from '../../injected/visualization-instance-processor';

const needsReviewRuleAnalyzerConfiguration: RuleAnalyzerConfiguration = {
    rules: ['aria-input-field-name', 'color-contrast', 'th-has-data-cells', 'link-in-text-block'],
    resultProcessor: (scanner: ScannerUtils) => scanner.getFailingInstances,
    telemetryProcessor: (telemetryFactory: TelemetryDataFactory) =>
        telemetryFactory.forNeedsReviewAnalyzerScan,
    key: AdHocTestkeys.NeedsReview,
    testType: VisualizationType.NeedsReview,
    analyzerMessageType: Messages.Visualizations.Common.ScanCompleted,
};

export const NeedsReviewAdHocVisualization: VisualizationConfiguration = {
    key: AdHocTestkeys.NeedsReview,
    testMode: TestMode.Adhoc,
    getTestView: props => (
        <AdhocIssuesTestView instancesSection={NeedsReviewInstancesSection} {...props} />
    ),
    getStoreData: data => data.adhoc.needsReview,
    enableTest: data => (data.enabled = true),
    disableTest: data => (data.enabled = false),
    getTestStatus: data => data.enabled,
    displayableData: {
        title: 'Needs review',
        subtitle: (
            <>
                <i>Needs review</i> automated checks will highlight elements that might have an
                accessibility issue. Examine each instance to determine if there is an accessibility
                issue.
            </>
        ),
        enableMessage: 'Running needs review checks...',
        toggleLabel: 'Show elements needing review',
        linkToDetailsViewText: 'List view and filtering',
    },
    launchPanelDisplayOrder: 6,
    adhocToolsPanelDisplayOrder: 6,
    getAnalyzer: provider =>
        provider.createRuleAnalyzerUnifiedScanForNeedsReview(needsReviewRuleAnalyzerConfiguration),
    getIdentifier: () => AdHocTestkeys.NeedsReview,
    visualizationInstanceProcessor: () => VisualizationInstanceProcessor.nullProcessor,
    getNotificationMessage: (selectorMap, key, warnings) =>
        getNotificationMessage(selectorMap, warnings),
    getDrawer: provider => provider.createHighlightBoxDrawer(),
    getSwitchToTargetTabOnScan: () => false,
    getInstanceIdentiferGenerator: () => generateUID,
    featureFlagToEnable: FeatureFlags.needsReview,
};
