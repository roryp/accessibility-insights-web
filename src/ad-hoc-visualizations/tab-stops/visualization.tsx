// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { AdHocTestkeys } from 'common/configs/adhoc-test-keys';
import { TestMode } from 'common/configs/test-mode';
import { VisualizationConfiguration } from 'common/configs/visualization-configuration';
import { Messages } from 'common/messages';
import { VisualizationType } from 'common/types/visualization-type';
import { generateUID } from 'common/uid-generator';
import { adhoc as content } from 'content/adhoc';
import { createHowToTest } from 'content/adhoc/tabstops/how-to-test';
import { AdhocStaticTestView } from 'DetailsView/components/adhoc-static-test-view';
import { FocusAnalyzerConfiguration } from 'injected/analyzers/analyzer';
import { VisualizationInstanceProcessor } from 'injected/visualization-instance-processor';
import * as React from 'react';

const { guidance, extraGuidance } = content.tabstops;

const tabStopVisualizationConfiguration: FocusAnalyzerConfiguration = {
    key: AdHocTestkeys.TabStops,
    testType: VisualizationType.TabStops,
    analyzerMessageType: Messages.Visualizations.Common.ScanCompleted,
    analyzerProgressMessageType: Messages.Visualizations.TabStops.TabbedElementAdded,
    analyzerTerminatedMessageType: Messages.Visualizations.TabStops.TerminateScan,
};

export const TabStopsAdHocVisualization: VisualizationConfiguration = {
    getTestView: props => (
        <AdhocStaticTestView content={createHowToTest(2)} guidance={extraGuidance} {...props} />
    ),
    key: AdHocTestkeys.TabStops,
    testMode: TestMode.Adhoc,
    getStoreData: data => data.adhoc.tabStops,
    enableTest: (data, _) => (data.enabled = true),
    disableTest: data => (data.enabled = false),
    getTestStatus: data => data.enabled,
    displayableData: {
        title: 'Tab stops',
        enableMessage: 'Start pressing Tab to start visualizing tab stops.',
        toggleLabel: 'Show tab stops',
        linkToDetailsViewText: 'How to test tab stops',
    },
    chromeCommand: '04_toggle-tabStops',
    launchPanelDisplayOrder: 4,
    adhocToolsPanelDisplayOrder: 5,
    analyzerProgressMessageType: Messages.Visualizations.TabStops.TabbedElementAdded,
    analyzerTerminatedMessageType: Messages.Visualizations.TabStops.TerminateScan,
    getAnalyzer: provider =>
        provider.createFocusTrackingAnalyzer(tabStopVisualizationConfiguration),
    getIdentifier: () => AdHocTestkeys.TabStops,
    visualizationInstanceProcessor: () => VisualizationInstanceProcessor.nullProcessor,
    getDrawer: provider => provider.createSVGDrawer(),
    getNotificationMessage: selectorMap => 'Start pressing Tab to start visualizing tab stops.',
    getSwitchToTargetTabOnScan: () => true,
    getInstanceIdentiferGenerator: () => generateUID,
    guidance,
};
