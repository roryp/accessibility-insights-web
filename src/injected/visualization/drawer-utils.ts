// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { ClientRectOffset } from '../client-utils';
export class DrawerUtils {
    private dom: Document;
    public clientWindowOffsetThreshold: number = 5;

    constructor(dom) {
        this.dom = dom;
    }

    public getContainerLeftOffset(offset: ClientRectOffset): number {
        return Math.max(this.clientWindowOffsetThreshold, offset.left);
    }

    public getContainerTopOffset(offset: ClientRectOffset): number {
        return Math.max(this.clientWindowOffsetThreshold, offset.top);
    }

    public getDocumentElement(): Document {
        return this.dom.ownerDocument || this.dom;
    }

    public getContainerWidth(
        offset: ClientRectOffset,
        doc: Document,
        elementBoundingClientRectWidth: number,
        bodyStyle: CSSStyleDeclaration,
        docStyle: CSSStyleDeclaration,
    ): number {
        const leftOffset = this.getContainerLeftOffset(offset);
        let containerWidth = elementBoundingClientRectWidth;

        if (
            leftOffset + containerWidth >=
            this.getDocumentWidth(doc, bodyStyle, docStyle) - this.clientWindowOffsetThreshold
        ) {
            containerWidth =
                this.getDocumentWidth(doc, bodyStyle, docStyle) -
                this.clientWindowOffsetThreshold -
                leftOffset;
        }

        return containerWidth;
    }

    public getContainerHeight(
        offset: ClientRectOffset,
        doc: Document,
        elementBoundingClientRectHeight: number,
        bodyStyle: CSSStyleDeclaration,
        docStyle: CSSStyleDeclaration,
    ): number {
        const topOffset = this.getContainerTopOffset(offset);
        let containerHeight = elementBoundingClientRectHeight;

        if (
            topOffset + containerHeight >=
            this.getDocumentHeight(doc, bodyStyle, docStyle) - this.clientWindowOffsetThreshold
        ) {
            containerHeight =
                this.getDocumentHeight(doc, bodyStyle, docStyle) -
                this.clientWindowOffsetThreshold -
                topOffset;
        }

        return containerHeight;
    }

    public isOutsideOfDocument(
        offset: ClientRectOffset,
        doc: Document,
        bodyStyle: CSSStyleDeclaration,
        docStyle: CSSStyleDeclaration,
    ): boolean {
        if (
            offset.left >=
            this.getDocumentWidth(doc, bodyStyle, docStyle) - this.clientWindowOffsetThreshold
        ) {
            return true;
        }

        if (
            offset.top >=
            this.getDocumentHeight(doc, bodyStyle, docStyle) - this.clientWindowOffsetThreshold
        ) {
            return true;
        }

        return false;
    }

    public getDocumentWidth(
        doc: Document,
        bodyStyle: CSSStyleDeclaration,
        docStyle: CSSStyleDeclaration,
    ): number {
        if (bodyStyle.overflowX === 'hidden' || docStyle.overflowX === 'hidden') {
            return doc.documentElement.clientWidth;
        }

        return Math.max(doc.body.scrollWidth, doc.documentElement.scrollWidth);
    }

    public getDocumentHeight(
        doc: Document,
        bodyStyle: CSSStyleDeclaration,
        docStyle: CSSStyleDeclaration,
    ): number {
        if (bodyStyle.overflowY === 'hidden' || docStyle.overflowY === 'hidden') {
            return doc.documentElement.clientHeight;
        }

        return Math.max(doc.body.scrollHeight, doc.documentElement.scrollHeight);
    }

    public static getBoundingClientRectIncludingChildren(element: Element): ClientRect {
        const rects: ClientRect[] = [];
        for (let i = 0; i < element.children.length; i++) {
            const boundingRect = element.children[i].getBoundingClientRect();
            if (boundingRect.height * boundingRect.width > 0) {
                rects.push(boundingRect);
            }
        }

        const initialRect = element.getBoundingClientRect();
        let left = initialRect.left;
        let top = initialRect.top;
        let right = initialRect.right;
        let bottom = initialRect.bottom;
        rects.forEach(r => {
            left = Math.min(left, r.left);
            top = Math.min(top, r.top);
            right = Math.max(right, r.right);
            bottom = Math.max(bottom, r.bottom);
        });

        return {
            left,
            x: left,
            y: top,
            top,
            bottom,
            right,
            width: right - left,
            height: bottom - top,
        };
    }
}
