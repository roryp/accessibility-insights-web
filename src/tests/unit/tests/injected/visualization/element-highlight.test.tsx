// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { shallow } from 'enzyme';
import * as React from 'react';
import { IMock, Mock } from 'typemoq';

import { FeatureFlagStoreData } from '../../../../../common/types/store-data/feature-flag-store-data';
import { ClientRectOffset, ClientUtils } from '../../../../../injected/client-utils';
import { RenderDialog } from '../../../../../injected/dialog-renderer';
import { HtmlElementAxeResults } from '../../../../../injected/scanner-utils';
import { DrawerUtils } from '../../../../../injected/visualization/drawer-utils';
import {
    ElementHighlight,
    ElementHighlightDeps,
    ElementHighlightProps,
} from '../../../../../injected/visualization/element-highlight';
import {
    BoxConfig,
    DrawerConfiguration,
    GetBoundingRect,
} from '../../../../../injected/visualization/formatter';

describe('ElementHighlight', () => {
    let deps: ElementHighlightDeps;
    let drawerConfiguration: DrawerConfiguration;
    let failureBoxConfigStub: BoxConfig;
    let textBoxConfigStub: BoxConfig;
    let props: ElementHighlightProps;
    let drawerUtilsMock: IMock<DrawerUtils>;
    let clientUtilsMock: IMock<ClientUtils>;
    let elementStub: Element;
    let elementResultStub: HtmlElementAxeResults;
    let bodyStyleStub: CSSStyleDeclaration;
    let docStyleStub: CSSStyleDeclaration;
    let featureFlagStoreDataStub: FeatureFlagStoreData;
    let dialogRenderMock: IMock<RenderDialog>;
    let getBoundingRect: IMock<GetBoundingRect>;
    let elementBoundingClientRectStub: ClientRect;
    let documentMock: IMock<Document>;
    let offsetStub: ClientRectOffset;

    beforeEach(() => {
        drawerUtilsMock = Mock.ofType(DrawerUtils);
        clientUtilsMock = Mock.ofType(ClientUtils);
        dialogRenderMock = Mock.ofType<RenderDialog>();
        getBoundingRect = Mock.ofType<GetBoundingRect>();
        documentMock = Mock.ofType<Document>();

        elementResultStub = {} as HtmlElementAxeResults;
        featureFlagStoreDataStub = {} as FeatureFlagStoreData;
        elementStub = {} as Element;
        elementBoundingClientRectStub = {
            bottom: 1,
            left: 2,
            height: 3,
            right: 4,
            top: 5,
            width: 6,
        };
        offsetStub = {
            left: 11,
            top: 22,
        };
        bodyStyleStub = {} as CSSStyleDeclaration;
        docStyleStub = {} as CSSStyleDeclaration;
        failureBoxConfigStub = {} as BoxConfig;
        textBoxConfigStub = {} as BoxConfig;

        drawerConfiguration = {
            showVisualization: true,
            borderColor: 'red',
            outlineStyle: 'solid',
            failureBoxConfig: failureBoxConfigStub,
            textBoxConfig: textBoxConfigStub,
            toolTip: 'some tool tip',
        };

        deps = {
            clientUtils: clientUtilsMock.object,
            drawerUtils: drawerUtilsMock.object,
        } as ElementHighlightDeps;

        props = {
            deps: deps,
            drawerConfig: drawerConfiguration,
            element: elementStub,
            elementResult: elementResultStub,
            bodyStyle: bodyStyleStub,
            docStyle: docStyleStub,
            featureFlagStoreData: featureFlagStoreDataStub,
            dialogRender: dialogRenderMock.object,
            getBoundingRect: getBoundingRect.object,
        };
    });

    test('renders null when drawerConfig showVisualization is false', () => {
        drawerConfiguration.showVisualization = false;
        const testObject = shallow(<ElementHighlight {...props} />);
        expect(testObject.getElement()).toMatchSnapshot();
    });

    test('renders null when element is outside of document', () => {
        getBoundingRect
            .setup(mock => mock(elementStub))
            .returns(() => elementBoundingClientRectStub);
        drawerUtilsMock.setup(mock => mock.getDocumentElement()).returns(() => documentMock.object);
        drawerUtilsMock
            .setup(mock =>
                mock.isOutsideOfDocument(
                    elementBoundingClientRectStub,
                    documentMock.object,
                    bodyStyleStub,
                    docStyleStub,
                ),
            )
            .returns(() => true);
        const testObject = shallow(<ElementHighlight {...props} />);
        expect(testObject.getElement()).toMatchSnapshot();
    });

    test('renders elements appropriately', () => {
        const styleTopStub = 111;
        const styleLeftStub = 222;
        const minWidthStub = 333;
        const minHeightStub = 444;

        getBoundingRect
            .setup(mock => mock(elementStub))
            .returns(() => elementBoundingClientRectStub);
        drawerUtilsMock.setup(mock => mock.getDocumentElement()).returns(() => documentMock.object);

        drawerUtilsMock
            .setup(mock =>
                mock.isOutsideOfDocument(
                    elementBoundingClientRectStub,
                    documentMock.object,
                    bodyStyleStub,
                    docStyleStub,
                ),
            )
            .returns(() => false);

        clientUtilsMock.setup(mock => mock.getOffset(elementStub)).returns(() => offsetStub);
        drawerUtilsMock
            .setup(mock => mock.getContainerTopOffset(offsetStub))
            .returns(() => styleTopStub);
        drawerUtilsMock
            .setup(mock => mock.getContainerLeftOffset(offsetStub))
            .returns(() => styleLeftStub);

        drawerUtilsMock
            .setup(mock =>
                mock.getContainerWidth(
                    offsetStub,
                    documentMock.object,
                    elementBoundingClientRectStub.width,
                    bodyStyleStub,
                    docStyleStub,
                ),
            )
            .returns(() => minWidthStub);

        drawerUtilsMock
            .setup(mock =>
                mock.getContainerHeight(
                    offsetStub,
                    documentMock.object,
                    elementBoundingClientRectStub.height,
                    bodyStyleStub,
                    docStyleStub,
                ),
            )
            .returns(() => minHeightStub);

        dialogRenderMock
            .setup(mock => mock(elementResultStub, featureFlagStoreDataStub))
            .verifiable();

        const testObject = shallow(<ElementHighlight {...props} />);
        const failureHighlightBox = testObject.findWhere(
            node => node.prop('onClickHandler') != null,
        );
        failureHighlightBox.prop('onClickHandler')();

        expect(testObject.getElement()).toMatchSnapshot();
        dialogRenderMock.verifyAll();
    });
});
