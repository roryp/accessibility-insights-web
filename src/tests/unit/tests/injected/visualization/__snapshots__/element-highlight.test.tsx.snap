// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ElementHighlight renders elements appropriately 1`] = `
<div
  className="insights-highlight-box"
  style={
    Object {
      "left": 222,
      "minHeight": 444,
      "minWidth": 333,
      "outlineColor": "red",
      "outlineStyle": "solid",
      "top": 111,
    }
  }
  title="some tool tip"
>
  <HighlightBox
    boxConfig={Object {}}
    className="text-label"
    deps={
      Object {
        "clientUtils": proxy {
          "___id": "BCDF5CE5-F0DF-40B7-8BA0-69DF395033C8",
          "scroll": undefined,
        },
        "drawerUtils": proxy {
          "___id": "BCDF5CE5-F0DF-40B7-8BA0-69DF395033C8",
          "clientWindowOffsetThreshold": 5,
          "dom": undefined,
        },
      }
    }
    drawerConfig={
      Object {
        "borderColor": "red",
        "failureBoxConfig": Object {},
        "outlineStyle": "solid",
        "showVisualization": true,
        "textBoxConfig": Object {},
        "toolTip": "some tool tip",
      }
    }
  />
  <HighlightBox
    boxConfig={Object {}}
    className="failure-label"
    deps={
      Object {
        "clientUtils": proxy {
          "___id": "BCDF5CE5-F0DF-40B7-8BA0-69DF395033C8",
          "scroll": undefined,
        },
        "drawerUtils": proxy {
          "___id": "BCDF5CE5-F0DF-40B7-8BA0-69DF395033C8",
          "clientWindowOffsetThreshold": 5,
          "dom": undefined,
        },
      }
    }
    drawerConfig={
      Object {
        "borderColor": "red",
        "failureBoxConfig": Object {},
        "outlineStyle": "solid",
        "showVisualization": true,
        "textBoxConfig": Object {},
        "toolTip": "some tool tip",
      }
    }
    onClickHandler={[Function]}
  />
</div>
`;

exports[`ElementHighlight renders null when drawerConfig showVisualization is false 1`] = `null`;

exports[`ElementHighlight renders null when element is outside of document 1`] = `null`;
